--------------
title: "Visualization for Starfish Discovery"
--------------

```{r setup, echo=FALSE}
# Authors: Jonah Cullen, Callum MacPhillamy, Mauricio Moldes, Alexis Norris, Meghana Ram, Marcus Chan, and Russel Santos

# Packages
library(shiny)
library(tidyverse)
library(ggExtra)
library(ggridges)
library(ggrepel)
library(ggpubfigs) # remotes::install_github("JLSteenwyk/ggpubfigs")
library(ggseqlogo)
library(hrbrthemes)
library(seqinr)
library(ggseqlogo)
library(cowplot)
library(treemapify)
library(mgcv)
library(shadowtext)
library(ggtree) # BiocManager::install("ggtree")
library(treeio)
library(ape)
library(phytools)
```

```{r input_testing, echo=FALSE, eval=FALSE}
inputs <- list(
  "benchmarkFile" = "", # starfishBenchmarkStats.txt
  "pairwiseFile" = "funTyr50_cap25_crp3_p1-512_activeFilt.clipkit.pid_byfam.txt",
  "taxonomicFile" = "tyrsPerGenome.txt",
  "speciesFile" = "tyrsPerSpecies.txt"
)
```

```{r}
ui <- fluidPage(
  
  # Application title
  titlePanel("Starfish Discovery"),
  
  # Sidebar layout with input and output definitions
  sidebarLayout(
    sidebarPanel(
      textInput("dir1", "Directory for Starfish Benchmarking:", 
                "ilikebigmobileelementsandicannotlie/benchmarking/"),
      textInput("dir2", "Directory for Tyr Pairwise Identities:", 
                "../../Figure3/tyr/"),
      textInput("dir3", "Directory for Tyr Taxonomic Distribution:", 
                "../../Figure3/tyr/"),
      textInput("dir4", "Directory for Elements per Genome:", 
                "../../Figure3/tyr/"),
      textInput("benchmarkFile", "Starfish Benchmarking File:", 
                "starfishBenchmarkStats.txt"),
      textInput("pairwiseFile", "Tyr Pairwise Identities File:", 
                "funTyr50_cap25_crp3_p1-512_activeFilt.clipkit.pid_byfam.txt"),
      textInput("taxonomicFile", "Tyr Taxonomic Distribution File:", 
                "tyrsPerGenome.txt"),
      textInput("speciesFile", "Elements per Genome File:", 
                "tyrsPerSpecies.txt")
    ),
    
    # Main panel for displaying outputs
    mainPanel(
      tabsetPanel(
        tabPanel("Starfish Benchmarking", plotOutput("starfish_plot")),
        tabPanel("Tyr Pairwise Identities", plotOutput("tyr_pairwise_plot")),
        tabPanel("Tyr Taxonomic Distribution", plotOutput("tyr_taxonomic_plot")),
        tabPanel("Elements per Genome", plotOutput("elements_per_genome_plot"))
      )
    )
  )
)

# Define server logic required to draw the plots
server <- function(input, output) {

  # Starfish Benchmarking plot
  # INPUT: starfishBenchmarkStats.txt
  output$starfish_plot <- renderPlot({
    
    read.delim(input$benchmarkFile) %>%
      mutate(module = fct_relevel(module, "captain", "boundary", "DR")) %>%
      ggplot() +
      geom_col(aes(x = benchmark, y = F1_score), fill = "darkgray", width = 0.75) + 
      geom_text(
        aes(x = benchmark, y = as.double(F1_score) + 0.025, label = F1_score), 
        angle = 0, size = 4
      ) +
      facet_grid(.~ module, scales = "free",  space = "free") +
      ggpubfigs::theme_grey() + 
      theme(
        text  = element_text(size = 15, family = "Helvetica"),
        axis.text.x = element_text(angle = 45)
      ) +
      labs(
        title = "Starfish performance on 42 manually annotated elements", 
        x = "Starfish module", y = "F1 score"
      )
    
  })

  ### Tyr Pairwise Identities plot
  ### INPUT: funTyr50_cap25_crp3_p1-512_activeFilt.clipkit.pid_byfam.txt
  output$tyr_pairwise_plot <- renderPlot({
    
    read.delim(input$pairwiseFile) %>%
      mutate(percentID = pid * 100) %>%
      dplyr::select(comparisonType, percentID) %>%
      ggplot %>%
      geom_violin(aes(x = fct_rev(comparisonType), y = percentID), fill = "darkgray") + 
      geom_boxplot(
        aes(x =  fct_rev(comparisonType), y = percentID), 
        fill = "white", width = 0.2, 
        outlier.size = 0, outlier.fill = NA, outlier.color = NA
      ) + 
      ggpubfigs::theme_grey() + 
      theme(text  = element_text(size = 18, family = "Helvetica")) +
      scale_x_discrete(labels = c("within family", "between family")) +
      labs(
        title = "Pairwise seqID of YR refs, from p1-512 alignment", 
        x = "Comparison type", y = "Pairwise sequence identity (%)"
      )
    
  })
  
  # Tyr Taxonomic Distribution plot
  # INPUT: tyrsPerGenome.txt
  output$tyr_taxonomic_plot <- renderPlot({

    classOrder <- c(
      "Pucciniomycotina", "Pucciniomycetes", "Ustilaginomycotina", "Ustilaginomycetes", 
      "Agaricomycetes", "Dacrymycetes", "Tremellomycetes", "Wallemiomycetes", "Pezizomycetes", 
      "Orbiliomycetes", "Eurotiomycetes", "Dothideomycetes", "Lecanoromycetes", "Leotiomycetes", 
      "Sordariomycetes", "Xylonomycetes", "Saccharomycotina", "Taphrinomycotina", 
      "Mucoromycota", "Glomeromycotina", "Mortierellomycotina", "Mucormycotina", 
      "Zoopagomycota", "Zoopagomycotina", "Entomophthoromycotina", "Kickxellomycotina", 
      "Chytridiomycota", "Blastocladiomycota", "Chytridiomycetes", "Monoblepharidomycetes", 
      "Neocallimastigomycetes", "Microsporidia", "Cryptomycota"
    )

    read.delim(input$taxonomicFile) %>%
      mutate(taxonID = factor(taxonID, levels = classOrder)) %>%
      dplyr::select(taxonID, tyrCount) %>%
      dplyr::filter(taxonID %in% classOrder) %>%
      ggplot() +
      geom_jitter(
        aes(x = fct_rev(taxonID), y = tyrCount), 
        width = 0.25, height = 0.25, shape = 21, fill = "gray", 
        color = "#3B3B3B", alpha = 0.4, size = 2
      ) +
      geom_boxplot(
        aes(x =  fct_rev(taxonID), y = tyrCount), 
        fill = "white", width = 0.75, size = 0.6, 
        outlier.size = 0, outlier.fill = NA, outlier.color = NA
      ) + 
      ggpubfigs::theme_grey() + 
      labs(
        title = "YR taxonomic distribution", 
        x = "Class ID", y = "YR count per genome"
      ) +
      scale_y_continuous(limits = c(0, 30)) +
      coord_flip() +
      theme(
        text  = element_text(size = 6, family = "Helvetica"),
        axis.text.y = element_text(hjust = 0)
      )
    
  })
  
  # Elements per Genome plot
  # INPUT: tyrsPerSpecies.txt
  output$elements_per_genome_plot <- renderPlot({

    tyrSpecies <- read.delim(input$speciesFile)
    elementSpecies <- tyrSpecies %>%
      group_by(taxonID) %>%
      tally(elementCount) %>%
      dplyr::filter(n > 0) %>%
      dplyr::filter(!grepl('sp\\.', taxonID)) %>%
      dplyr::select(taxonID)
    
    tyrSpecies %>%
      dplyr::filter(taxonID %in% elementSpecies$taxonID) %>%
      ggplot() +
      geom_jitter(
        aes(x = taxonCount, y = elementCount), 
        shape = 21, fill = "gray", color = "#3B3B3B", alpha = 0.4, size = 2
      ) +
      ggpubfigs::theme_grey() + 
      labs(
        x = "Genomes per species", y = "Elements per genome",
        title = "Elements per genome as a function of genomes per species"
      ) +
      theme(
        text  = element_text(size = 6, family = "Helvetica"),
        axis.text.y = element_text(hjust = 0)
      )
    
  })
}

# Run the application 
shinyApp(ui = ui, server = server)
```

